[WIP] Issue/gl fixes

Lord-Kamina opened the pull request:
<!--
PLEASE READ THIS MESSAGE.

HOW TO WRITE A GOOD PULL REQUEST?

- Make it small.
- Do only one thing.
- Avoid re-formatting.
- Make sure the code builds.
- Write useful descriptions and titles.
- Address review comments in terms of additional commits.
- Do not amend/squash existing ones unless the PR is trivial.
- Read the contributing guide: https://github.com/performous/performous/blob/master/.github/CONTRIBUTING.md.

-->

### What does this PR do?

- Does away with most glmath.hh in favor of using glm, which is an actively maintained GL Math library.
- Fixes the way shaders are compiled and linked so the geometry shader is now actually optional, instead of having a hacky pass-through mode that always runs.
- Fixed some structs and other issues in the shaders, rewrote the geometry shader (fixes #174!)
- Rewrote ShaderMap in order to use STL containers and unique_ptr instead of boost:ptr_map.

<!-- A brief description of the change being made with this pull request. -->

Shit was broken... Like, really.

### Closes Issue(s)
* closes #174 
<!-- List here all the issues closed by this pull request. -->

### Motivation
Seriously?
<!-- What inspired you to submit this pull request? -->


### More

- [ ] Added/updated documentation

### Additional Notes

<!-- Anything else we should know when reviewing? -->
Note: This PR is WIP because,

a) Currently it's just a duplicate of the win-master branch; and that branch contains commits that revert previous PRs, in order to make it build.
b) There's still some work to be done in the way we're handling GL on the C++ side of things.

note: This PR integrates also my previous PR with Unicode fixes so we probably won't want to merge until that one has been merged first anyway.

Lord-Kamina commented:
Can you guys take this one for a spin? I think I'm done with this (Ideally unicode PR should be merged first though since this one also has the changes from that one)

Report with suggestions/questions/bugs and/or comments on performance, please.

nieknooijens commented:
I think I have to take my time for this 36-commit-long monster. can't review it at work in my break. :worried:

Lord-Kamina commented:
Sure, just saying it's not really a WIP anymore.

nieknooijens commented:
ok I have 2 complaints.

first: add set(PROJECT_VERSION 1.0) to ced/CmakeList.txt
cmake was complaining

second.. oh well:
![schermafdruk van 2018-08-21 20-27-39](https://user-images.githubusercontent.com/2794331/44421450-d0116d00-a580-11e8-8ad0-4fcb1c1b2056.png)

I have no image at all... the game works, I hear music and I can push buttons to control it, but no image...

Lord-Kamina commented:
The "ced" issue is technically an issue with the unicode PR but ok. 

Re: the image issue, I'll take a look and make sure I didn't omit any commits or anything. What GPU/Distro/Version you on?

nieknooijens commented:
![schermafdruk van 2018-08-21 20-34-27](https://user-images.githubusercontent.com/2794331/44421696-9d1ba900-a581-11e8-8a42-3e3477fb1bd1.png)

Lord-Kamina commented:
Please add the config.xml also.

Finally, can you play around with manually editing the stereo3d settings to different values (on fullscreen ideally) and see if there's any change?

nieknooijens commented:
[config.xml.txt](https://github.com/performous/performous/files/2307592/config.xml.txt)

made it txt to make github accept it

Lord-Kamina commented:
Anything out of the ordinary in debug log?

Lord-Kamina commented:
I've made a bionic install and reproduced your issue. I see nothing in the log or anywhere. I'll see if I can find the issue.

nieknooijens commented:
nothing in the log, other than "you have UTF-8 issues with pokémon songs" which has been true for a long time now...

Lord-Kamina commented:
I have pinpointed the issue, it's caused by libglm 0.9.9.0 (windows and mac were on 0.9.8.x); I don't think it's a bug in the library but rather the update removed deprecated or wrong things that might be highlighting issues on our code.

Lord-Kamina commented:
@nieknooijens Fixed now.

In case you're interested, here's the cause: In our old code with the in-house glm, we initialized all matices as identity; when I changed to libglm I realized there's no identity function, since they were default-constructed to identity. libglm 0.9.9.0 removed the default constructor, which resulted in this error. To fix it, I had to expressly initialize vector types to 0.0f and matrix types to 1.0f, which is how an identity matrix is built according to the GLSL specification.

nieknooijens commented:
so basically we were using uninitialized memory?
I'll try checking it out tonight and otherwise tomorrow.

nieknooijens commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "ced"]
+	path = ced
+	url = https://Lord-Kamina@github.com/performous/compact_enc_det
```
make this link universal please, when I wanted to commit a fix it started to ask for your password.

Lord-Kamina replied:
Hahahaha hadn't even noticed.

nieknooijens commented on data/config/schema.xml:
```diff
@@ -94,6 +94,10 @@ to save the current settings to XML.
 	</entry>
+	<entry name="game/case-sorting" type="bool" value="false">
+		<short>Case-sensitive sorting.</short>
+		<long>Make a difference between uppercase and lowercase letters when sorting.</long>
```
nice one!

Lord-Kamina commented:
Will fix the .gitmodules and merge tonight then, unless @Baklap4 wants to weigh-in first?

Baklap4 commented:
I haven't had time to review yet, will do by tonight as i have some time free now 👍

Baklap4 commented:
I think it'd be wise to rebase this upon current master to remove the overlapping commits

Lord-Kamina commented:
@Baklap4 things I wouldn't wish upon my worst enemy: Solving merge conflicts from a phone. Let's hope I didn't break anything 😂

Baklap4 commented:
Hmm if i'm right you merged master into this branch (resulting in conflicts currently)
If you rebase it (`git checkout master && git pull && git checkout issue/gl_fixes && git rebase -i master`) you'll still have to solve the conflicts, but will remove the commits already made on master disappear from this PR. Resulting in a force push being needed since the local branch and remote branch are diversed. (`git push -f`)

Also this branch still contains the reverted commits (webserver + cache)

Lord-Kamina commented:
The reverted commits are reverted again. It's a double negation i.e. It just looks ugly. I haven't manually merged, rather just solved conflicts caused by the gitmodules change.

Lord-Kamina commented:
@Baklap4 I re-merged and rebased stuff to correct some commit messages and remove the reverted and reverted reverted commits. Please make sure I didn't unwittingly break anything before eventual merge?

Baklap4 commented:
I'll try it out tonight to make sure it doesn't break stuff ;) Thanks for rebasing!

Lord-Kamina commented:
> [32a37fc](https://github.com/performous/performous/commit/32a37fc7949e0c6d478adb5df765a3c9fba78e74) is doing some weird stuff with Webserver code

Kinda sorta.  It's because I originally began work on this branch on the parallel win-master branch (because the motivation was to fix the Windows rendering glitches) where I had reverted the switch to CppRest in order to be able to build.

dos1 commented:
OK, after rechecking, seems this branch works for me and fixes the segfault from #401 :)

However, window resize is now broken - the content inside always stays at the same, initial size.

Lord-Kamina commented:
> OK, after rechecking, seems this branch works for me and fixes the segfault from #401 :)
> 
> However, window resize is now broken - the content inside always stays at the same, initial size.

Can you try tracing the resize issue to see whether I made a mistake in my code (quite possible)

Lord-Kamina commented:
@baklap4 can you see whether you have any issues resizing? I'm fairly sure I tested it without issue but I DID mess with resizing code a bit so I might have missed something.

Baklap4 commented:
i can resize without problems however i can resize to a smaller window as performous can show.. resulting in images like this: 
![image](https://user-images.githubusercontent.com/3918517/46687665-b3e29180-cbfb-11e8-9259-1e552cbcbfd5.png)

maybe let's set a min width and height to which max may be downsized?

Lord-Kamina commented:
What happens if after resizing, you toggle fullscreen?

Most likely I made a mistake. Will take a look tonight/tomorrow

Baklap4 commented:
If you resized your window way smaller and go full screen it goes full screen and looking alright. If you disable fullscreen again it'll reset your window to contain the default resolution. (which is good)

Lord-Kamina commented:
Resizing should be fixed now.

Baklap4 commented:
This branch currently has conflicting files..

Lord-Kamina commented:
I'll check tonight since github seemingly won't allow me to solve the conflicts via its UI

Lord-Kamina commented:
Just noticed I'd broken it with the merge the other day, I must have been very tired because I'd resolved the conflict as SWFormat and AVResample 😂

Lord-Kamina commented:
I have re-rebased this monster; theoretically the cpprestsdk reverts should be really gone and not just disguised now.
